control/man/astroinstrument.1
.TH ASTROINSTRUMENT 1 "Sept 2014"
.SH NAME
astroinstrument \- Instrument configuration tool

.SH SYNOPSIS
.B astroinstrument
[
.I options
] 
list

.B astroinstrument
[
.I options
] 
help

.B astroinstrument
[
.I options
] 
{ add | show | remove } <name>

.B astroinstrument
[
.I options
] 
.I name
add
.I cname
[
.I attr=value
]

.B astroinstrument
[
.I options
] 
.I name 
{ show | remove }
.I type
[
.I attr=value
]

.SH DESCRIPTION
An instrument is a named collection of astrophotography devices as managed
by the astrophotography tools. The purpose of have a simple handle on the
configuration used in an astrophotography project.
Usually an instrument includes a least a camera,
a CCD chip and a cooler for the CCD for a camera with cooled chips.
More complex instruments might include  filter wheel, an adaptive
optics system or a mount. 
The 
.BR astroinstrument (1)
is used to configure, update and view instrument configurations. 
To build an instrument, one first adds the instrument name using
the 
.B add
.I name
command, and then uses the 
.I name
.B add 
.I component
command to add components to the instrument.

There are three kinds of components that can be attached to an instrument.
.TP
direct component
A direct component describes the device directly through the device name
URI and a unit id.
.TP
mapped component
Mapped components are useful for situations where the device may change 
the device name when plugged in. In those cases, the
.BR devicemapper (1)
program can be used to create a stable name for the device. By only
adapting the device mapping, configurations refering to the stable name
will always get the correct name. Mapped components refer to the
device mapper name.
.TP
derived component
This kind of component is used when the component is really part of
some other part. This is most commonly used for CCDs, which are usually
part of the camera. 

.SH CONFIGURATION COMMANDS
.TP
.B help
Display a short help message.
.TP
.B list
This shows a list of instrument
names together with a comma separated list of components that are
attached to an instrument.
.TP
.BI add \ name
Add a new instrument named 
.IR name .
.TP
.BI show \ name
Show details of the instrument named
.IR name .
.TP
.BI remove \ name
Remove an instrument from the configuration database.
.TP
.IB name \ add \ type\ [\ attr=value\ ] 
Add a component of type 
.I type
to the instrument named
.IR name .
Valid component types are 
.IR adaptiveoptics , 
.IR camera ,
.IR ccd ,
.IR cooler ,
.IR filterwheel ,
.IR focuser ,
.IR guiderport ,
.I module 
and
.IR mount .
For valid attribute value pairs see the section ATTRIBUTES below.
.TP
.IB name \ update \ type\ [\ attr=value\ ] 
Update the component of type 
.I type
in the instrument named
.I name 
using the attribute value pairs.
For valid attribute value pairs see the section ATTRIBUTES below.
.TP
.B name \ {\ show\ |\ remove\ } \ type
Show or remove the component of type
.I type
in the instrument named
.IR name .

.SH ATTRIBUTES
The set of valid attributes depends on the kind of component being
manipulated. 
.TP
kind
The kind of the component to be created. Valid values are
.IR direct ,
.I mapped
and
.IR derived .
Default value is 
.IR direct ,
i.e. if nothing is specified, a direct component will be created.
.TP
unit
Defines the unit number for devices that support this concept.
Default value is 0.
.TP
device
The string form of the device url. Only for direct components, and
required for them.
.TP
name
Defines the device mapper name to use to access the real device.
Only for mapped components, and required for them.
.TP
from
Defines the type of component this component is derived from. Only
for derived components, and required for them.
.TP
servername
Name of the server on which this device is accessible. Only for direct
components. Derived components get their servername from the component
they are derived from. Mapped components get the server name from the
device mapping.

.SH OPTIONS
.TP
.B \-\-help,\-h
Display a help message.
.TP
.B \-\-debug,\-d
Increase the debug level
.TP
.BI \-\-config= config.db ,\-c \ config.db
The configuration is stored in a sqlite3 database file. This option
can be used to use a different database file than the default
.IR ~/.astrophoto.db .

.SH FILES
.I ~/.astrophoto.db

.SH AUTHOR
Andreas Mueller


control/tools/astroinstrument.cpp
/*
 * astroinstrument.cpp -- astrophotography instrument construction
 *
 * (c) 2014 Prof Dr Andreas Mueller, Hochschule Rapperswil
 */
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <includes.h>
#include <AstroDebug.h>
#include <AstroConfig.h>
#include <AstroUtils.h>
#include <AstroPersistence.h>
#include <InstrumentTables.h>
#include <algorithm>

using namespace astro;
using namespace astro::config;
using namespace astro::persistence;

namespace astro {
namespace app {
namespace instrument {

/**
 * \brief Table of options for the astroinstrument
 */
static struct option	longopts[] = {
/* name		argument?		int*		int */
{ "config",	required_argument,	NULL,		'c' }, /* 0 */
{ "debug",	no_argument,		NULL,		'd' }, /* 1 */
{ "help",	no_argument,		NULL,		'h' }, /* 2 */
{ NULL,		0,			NULL,		0   } 
};

/**
 * \brief Usage function for the 
 */
void	usage(const std::string& progname) {
	std::string	prg = std::string("    ") + Path(progname).basename();
	std::cout << "Usage:" << std::endl;
	std::cout << std::endl;
	std::cout << prg << " [ options ] list" << std::endl;
	std::cout << prg << " [ options ] { add | show | remove } <name>";
	std::cout << std::endl;
	std::cout << prg << " [ options ] <name> add <cname> [ attr=value ]";
	std::cout << std::endl;
	std::cout << prg << " [ options ] <name> { show | remove } <type>";
	std::cout << std::endl;
	std::cout << prg << " [ options ] <name> update <type> "
		"[ attr=value ]";
	std::cout << std::endl;
	std::cout << prg << " [ options ] help" << std::endl;
	std::cout << std::endl;
	std::cout << "The following attributes are known (in brackets the "
		"kinds of components" << std::endl;
	std::cout << "supporting this attribute, the letters mean: d=derived, "
		"m=mapped, r=derived):" << std::endl;
	std::cout << "  unit=<u>      unit number of a device [all]";
	std::cout << std::endl;
	std::cout << "  device=<d>    device name of a direct component [d]";
	std::cout << std::endl;
	std::cout << "  name=<n>      name of a mapped component [m]" << std::endl;
	std::cout << "  kind=<k>      component kind (direct, mapped, derived) [all]";
	std::cout << std::endl;
	std::cout << "  from=<f>      the device type from which this component"
		" is derived" << std::endl;
	std::cout << "  server=<s>    component is accessible on server <s>";
	std::cout << std::endl;
	std::cout << std::endl;
	std::cout << "Options:" << std::endl;
	std::cout << " -c,--config=<cfg>   use configuraton file <cfg> instead "
			"of the default";
	std::cout << std::endl;
	std::cout << " -d,--debug          increase debug level" << std::endl;
	std::cout << " -h,--help           show help message and exit";
	std::cout << std::endl;
}

/**
 * \brief display a help message
 */
int	cmd_help() {
	usage("astroinstrument");
	return EXIT_SUCCESS;
}

/**
 * \brief List all known instruments
 */
int	cmd_list() {
	std::list<InstrumentPtr>	instruments
		= Configuration::get()->listinstruments();
	std::for_each(instruments.begin(), instruments.end(),
		[](InstrumentPtr instrument) {
			std::cout << instrument->toString() << std::endl;
		}
	);
	return EXIT_SUCCESS;
}

/**
 * \brief add an instrument
 *
 * Adding an instrument needs only the name of the instrument
 */
int	cmd_add(const std::string& instrumentname,
		const std::vector<std::string>& /* arguments */) {
	debug(LOG_DEBUG, DEBUG_LOG, 0, "adding instrument '%s'",
		instrumentname.c_str());
	ConfigurationPtr	config = Configuration::get();
	debug(LOG_DEBUG, DEBUG_LOG, 0, "creating new instrument");
	InstrumentPtr	instrument(new Instrument(config->database(),
				instrumentname));
	debug(LOG_DEBUG, DEBUG_LOG, 0, "instrument created");
	return EXIT_SUCCESS;
}

/**
 * \brief show an instrument
 */
int	cmd_show(const std::string& instrumentname,
		const std::vector<std::string>& /* arguments */) {
	InstrumentPtr	instrument
		= Configuration::get()->instrument(instrumentname);
	std::cout << instrument->name();
	std::list<DeviceName::device_type>	types
		= instrument->component_types();
	std::cout << " has " << types.size() << " components" << std::endl;
	for (auto ptr = types.begin(); ptr != types.end(); ptr++) {
		InstrumentComponentPtr	component = instrument->component(*ptr);
		std::cout << component->toString() << std::endl;
	}
	return EXIT_SUCCESS;
}

/**
 * \brief Remove an instrument
 */
int	cmd_remove(const std::string& instrumentname,
		const std::vector<std::string>& /* arguments */) {
	Configuration::get()->removeInstrument(instrumentname);
	return EXIT_SUCCESS;
}

/**
 * \brief add a component to an instrument
 *
 * Adding an instrument uses the command format
 *
 *     INSTR add cname attr=value
 *
 * INSTR is the name of the instrument, cname is the device type of the
 * component to be added. Some of the valid device types are camera,
 * ccd, adaptiveoptics, cooler, filterwheel, ... 
 * The following attribute value pairs are unit=<u>, kind=<t> (required),
 * device=<d>, name=<n>, from=<f>
 */
int	cmd_component_add(const std::string& instrumentname,
		const std::string& componenttype,
		const std::vector<std::string>& arguments) {
	// get the instrument to be changed
	ConfigurationPtr	config = Configuration::get();
	InstrumentPtr	instrument = config->instrument(instrumentname);

	// interpret the component type
	debug(LOG_DEBUG, DEBUG_LOG, 0, "component type: %s",
		componenttype.c_str());
	DeviceName::device_type	type
		= InstrumentComponentTableAdapter::type(componenttype);

	// build the component
	InstrumentComponentPtr	component(NULL);

	// parse the arguments
	AttributeValuePairs	av(arguments, 3);

	// check for the type
	InstrumentComponent::component_t 	ctype
		= InstrumentComponent::direct;
	if (av.has("kind")) {
		std::string	t = av("kind");
		ctype = InstrumentComponentTableAdapter::component_type(t);
	}
	debug(LOG_DEBUG, DEBUG_LOG, 0, "the component type is %s", 
		InstrumentComponentTableAdapter::component_type(ctype).c_str());;

	// get the unit number, as all types of components have this attribute
	int	unit = 0;
	if (av.has("unit")) {
		unit = std::stoi(av("unit"));
	}
	std::string	servername;
	if (av.has("server")) {
		servername = av("server");
	}

	// now use different code for the various types of component classes
	switch (ctype) {
	case InstrumentComponent::direct:
		component = InstrumentComponentPtr(
			new InstrumentComponentDirect(type,
				DeviceName(av("device")), unit, servername));
		break;
	case InstrumentComponent::mapped:
		// construct a mapped component
		component = InstrumentComponentPtr(
			new InstrumentComponentMapped(type,
				config->database(), av("name")));
		break;
	case InstrumentComponent::derived:
		// construct a derived compont, where from attribute contains
		// the type of the component this component is derived from
		component = InstrumentComponentPtr(
			new InstrumentComponentDerived(type, *instrument,
				InstrumentComponentTableAdapter::type(av("from")),
				unit));
		break;
	}

	// add the component to the instrument
	instrument->add(component);

	// persist in the database
	config->database()->begin("addcomponent");
	try {
		config->removeInstrument(instrumentname);
		config->addInstrument(instrument);
		config->database()->commit("addcomponent");
	} catch (...) {
		config->database()->rollback("addcomponent");
	}
	return EXIT_SUCCESS;
}

/**
 * \brief Update a component of an instrument
 */
int	cmd_component_update(const std::string& instrumentname,
		const std::string& componenttype,
		const std::vector<std::string>& arguments) {
	// get the component to change
	ConfigurationPtr	config = Configuration::get();
	InstrumentPtr	instrument = config->instrument(instrumentname);
	InstrumentComponentPtr	component = instrument->component(
		InstrumentComponentTableAdapter::type(componenttype));

	// get the attribute value types from the remaining arguments
	AttributeValuePairs	av(arguments, 3);

	// check for the type
	if (av.has("kind")) {
		InstrumentComponent::component_t	ctype
			= InstrumentComponentTableAdapter::component_type(
				av("kind"));
		if (ctype != component->component_type()) {
			throw std::runtime_error("cannot change type, delete "
				"and add component of new type");
		}
	}

	// XXX modify the component
	if (av.has("unit")) {
		component->unit(std::stoi(av("unit")));
	}
	switch (component->component_type()) {
	case InstrumentComponent::direct:
		if (av.has("device")) {
			component->name(av("device"));
		}
		if (av.has("server")) {
			debug(LOG_DEBUG, DEBUG_LOG, 0, "update servername to"
				" '%s'", av("server").c_str());
			dynamic_cast<InstrumentComponentDirect *>(&*component)
				->servername(av("server"));
		}
		break;
	case InstrumentComponent::mapped:
		if (av.has("name")) {
			component->name(av("name"));
		}
		break;
	case InstrumentComponent::derived:
		if (av.has("from")) {
			component->name(av("from"));
		}
		break;
	}

	// persist in the database
	config->database()->begin("updatecomponent");
	try {
		config->removeInstrument(instrumentname);
		config->addInstrument(instrument);
		config->database()->commit("updatecomponent");
	} catch (...) {
		config->database()->rollback("updatecomponent");
		throw;
	}
	return EXIT_SUCCESS;
}

/**
 * \brief Show the details of a component
 */
int	cmd_component_show(const std::string& instrumentname,
		const std::string& componenttype,
		const std::vector<std::string>& /* arguments */) {
	InstrumentPtr	instrument
		= Configuration::get()->instrument(instrumentname);
	InstrumentComponentPtr	component = instrument->component(
		InstrumentComponentTableAdapter::type(componenttype));
	std::cout << componenttype << " component of instrument ";
	std::cout << instrumentname  << ":" << std::endl;
	std::cout << component->toString() << std::endl;
	return EXIT_SUCCESS;
}

/**
 * \brief Remove a component from from an instrument
 */
int	cmd_component_remove(const std::string& instrumentname,
		const std::string& componenttype,
		const std::vector<std::string>& /* arguments */) {
	ConfigurationPtr	config = Configuration::get();
	InstrumentPtr	instrument
		= config->instrument(instrumentname);
	instrument->remove(
		InstrumentComponentTableAdapter::type(componenttype));
	config->database()->begin("removecomponent");
	try {
		config->removeInstrument(instrumentname);
		config->addInstrument(instrument);
		config->database()->commit("removecomponent");
	} catch (...) {
		config->database()->rollback("removecomponent");
		throw;
	}
	return EXIT_SUCCESS;
}

/**
 * \brief Interpret the various subcommands that astroinstrument implements
 */
int	commands(const std::vector<std::string>& arguments) {
	if (arguments.size() < 1) {
		throw std::runtime_error("not enought arguments");
	}
	// generic list command
	if (arguments[0] == "list") {
		return cmd_list();
	}
	if (arguments[0] == "help") {
		return cmd_help();
	}

	// instrument specific commands
	if (arguments.size() < 2) {
		throw std::runtime_error("not enough arguments");
	}
	std::string	instrumentname = arguments[1];
	debug(LOG_DEBUG, DEBUG_LOG, 0, "instrumentname = %s",
		instrumentname.c_str());
	if (arguments[0] == "add") {
		return cmd_add(instrumentname, arguments);
	}
	if (arguments[0] == "show") {
		return cmd_show(instrumentname, arguments);
	}
	if (arguments[0] == "remove") {
		return cmd_remove(instrumentname, arguments);
	}
	// component commands start here
	instrumentname = arguments[0];
	debug(LOG_DEBUG, DEBUG_LOG, 0, "instrumentname = %s",
		instrumentname.c_str());
	if (arguments.size() < 3) {
		throw std::runtime_error("not enough arguments");
	}
	std::string	componenttype = arguments[2];
	std::string	cmd = arguments[1];
	debug(LOG_DEBUG, DEBUG_LOG, 0, "command = %s", cmd.c_str());
	if (cmd == "add") {
		return cmd_component_add(instrumentname, componenttype,
			arguments);
	}
	if (cmd == "update") {
		return cmd_component_update(instrumentname, componenttype,
			arguments);
	}
	if (cmd == "remove") {
		return cmd_component_remove(instrumentname, componenttype,
			arguments);
	}
	if (cmd == "show") {
		return cmd_component_show(instrumentname, componenttype,
			arguments);
	}
	throw std::runtime_error("unknown command");
}

/**
 * \brief main method of the astroinstrument command
 */
int	main(int argc, char *argv[]) {
	int	c;
	int	longindex;
	while (EOF != (c = getopt_long(argc, argv, "c:dh", longopts,
		&longindex))) {
		switch (c) {
		case 'c':
			Configuration::set_default(std::string(optarg));
			break;
		case 'd':
			debuglevel = LOG_DEBUG;
			break;
		case 'h':
			usage(argv[0]);
			return EXIT_SUCCESS;
		case 1:
			switch (longindex) {
			}
			break;
		}
	}

	// remaining arguments are 
	std::vector<std::string>	arguments;
	while (optind < argc) {
		arguments.push_back(argv[optind++]);
	}

	// decide which command is requested
	return commands(arguments);
}

} // namespace instrument
} // namespace app
} // namespace astro

int	main(int argc, char *argv[]) {
	return astro::main_function<astro::app::instrument::main>(argc, argv);
}
